File: submit_rows.py
#!/usr/bin/env python3
"""Submit parsed rows JSON (stdin) to Notion."""

import difflib
import json
import os
import re
import sys
from datetime import date, datetime

import requests

from penny_pal_bot.expense_signs import normalize_rows as normalize_signed_rows, parse_amount

NOTION_TOKEN = os.getenv('NOTION_TOKEN')
if not NOTION_TOKEN:
    print('NOTION_TOKEN missing', file=sys.stderr)
    sys.exit(2)

if len(sys.argv) < 2:
    print('Usage: submit_rows.py <target_db_id>', file=sys.stderr)
    sys.exit(2)

target_db = sys.argv[1]
rows = json.load(sys.stdin)

headers = {
    'Authorization': f'Bearer {NOTION_TOKEN}',
    'Notion-Version': '2022-06-28',
    'Content-Type': 'application/json',
}
BASE_URL = 'https://api.notion.com/v1'


def normalize_text(value):
    return re.sub(r'[^a-z0-9]+', '', str(value or '').lower())


def normalize_date(value):
    if not value:
        return None
    s = str(value).strip()
    if not s:
        return None
    m = re.match(r'^(\d{4}-\d{2}-\d{2})(?:[T ].*)?$', s)
    if m:
        return m.group(1)

    today = date.today()
    formats = [
        '%d %b %Y', '%d %B %Y', '%d %b', '%d %B',
        '%b %d %Y', '%B %d %Y', '%b %d', '%B %d',
        '%d/%m/%Y', '%d/%m/%y', '%d/%m', '%d-%m-%Y', '%d-%m-%y', '%d-%m',
    ]
    for fmt in formats:
        try:
            dt = datetime.strptime(s, fmt)
            if '%Y' not in fmt and '%y' not in fmt:
                dt = dt.replace(year=today.year)
                if dt.date() > today and (dt.date() - today).days > 30:
                    dt = dt.replace(year=today.year - 1)
            return dt.date().isoformat()
        except ValueError:
            pass
    return None


def notion_get(path):
    resp = requests.get(f'{BASE_URL}{path}', headers=headers, timeout=30)
    if resp.status_code != 200:
        print('Notion GET failed', resp.status_code, resp.text, file=sys.stderr)
        sys.exit(3)
    return resp.json()


def notion_post(path, payload):
    return requests.post(f'{BASE_URL}{path}', headers=headers, json=payload, timeout=30)


def get_database_schema(db_id):
    return notion_get(f'/databases/{db_id}')


def query_database(db_id):
    results = []
    payload = {'page_size': 100}
    while True:
        resp = notion_post(f'/databases/{db_id}/query', payload)
        if resp.status_code != 200:
            print('Notion query failed', resp.status_code, resp.text, file=sys.stderr)
            sys.exit(3)
        data = resp.json()
        results.extend(data.get('results', []))
        if not data.get('has_more'):
            break
        payload['start_cursor'] = data.get('next_cursor')
    return results


def get_property_names(schema):
    props = schema.get('properties', {})
    found = {'title': None, 'amount': None, 'date': None, 'category': None, 'remarks': None}
    for name, prop in props.items():
        prop_type = prop.get('type')
        lowered = name.lower()
        if prop_type == 'title' and (found['title'] is None or lowered == 'item'):
            found['title'] = name
        elif prop_type == 'number' and (found['amount'] is None or lowered == 'amount'):
            found['amount'] = name
        elif prop_type == 'date' and (found['date'] is None or lowered == 'date'):
            found['date'] = name
        elif prop_type == 'relation' and (found['category'] is None or lowered == 'category'):
            found['category'] = name
        elif prop_type == 'rich_text' and (found['remarks'] is None or lowered == 'remarks'):
            found['remarks'] = name
    return found


def extract_title_from_page(page):
    props = page.get('properties', {})
    for prop in props.values():
        if prop.get('type') == 'title':
            parts = []
            for chunk in prop.get('title', []):
                parts.append(chunk.get('plain_text') or chunk.get('text', {}).get('content', ''))
            return ''.join(parts).strip()
    return ''


def get_relation_options(schema, prop_names):
    category_prop = prop_names.get('category')
    if not category_prop:
        return []
    relation_info = schema.get('properties', {}).get(category_prop, {}).get('relation', {})
    relation_db_id = relation_info.get('database_id')
    if not relation_db_id:
        return []
    pages = query_database(relation_db_id)
    options = []
    for page in pages:
        title = extract_title_from_page(page)
        if title:
            options.append({'id': page.get('id'), 'title': title})
    return options


def resolve_relation_id(raw_category, relation_options):
    if not raw_category or not relation_options:
        return None
    raw = str(raw_category).strip()
    raw_lower = raw.lower()
    normalized_title_map = {normalize_text(opt['title']): opt['id'] for opt in relation_options}
    aliases = [
        ('transport', 'Transport'), ('bus', 'Transport'), ('mrt', 'Transport'), ('grab', 'Transport'),
        ('taxi', 'Transport'), ('train', 'Transport'), ('food', 'Food'), ('grocer', 'Food'),
        ('shopping', 'Lifestyle'), ('lifestyle', 'Lifestyle'), ('software', 'Lifestyle'),
        ('subscription', 'Lifestyle'), ('invest', 'Investments'), ('goal', 'Goals'),
        ('saving', 'Savings'), ('tith', 'Tithing'), ('buffer', 'Buffer'),
    ]
    for needle, canonical in aliases:
        if needle in raw_lower:
            match_id = normalized_title_map.get(normalize_text(canonical))
            if match_id:
                return match_id
    raw_norm = normalize_text(raw)
    if raw_norm in normalized_title_map:
        return normalized_title_map[raw_norm]
    for opt in relation_options:
        opt_lower = opt['title'].lower()
        if raw_lower in opt_lower or opt_lower in raw_lower:
            return opt['id']
    close = difflib.get_close_matches(raw_norm, list(normalized_title_map.keys()), n=1, cutoff=0.55)
    if close:
        return normalized_title_map[close[0]]
    return None


def main():
    schema = get_database_schema(target_db)
    prop_names = get_property_names(schema)
    relation_options = get_relation_options(schema, prop_names)
    if not prop_names.get('title'):
        print('Could not find title property in target database schema', file=sys.stderr)
        sys.exit(3)

    signed_rows = normalize_signed_rows(rows)
    for row in signed_rows:
        props = {}
        item = row.get('item') or row.get('name') or 'Unknown'
        amt = row.get('amount')
        raw_date = row.get('date')
        cat = row.get('category')
        num = parse_amount(amt)
        props[prop_names['title']] = {'title': [{'text': {'content': str(item)}}]}
        if num is not None and prop_names.get('amount'):
            props[prop_names['amount']] = {'number': num}
        iso_date = normalize_date(raw_date)
        if iso_date and prop_names.get('date'):
            props[prop_names['date']] = {'date': {'start': iso_date}}
        elif raw_date:
            print(f'Invalid date after normalization: {raw_date!r}', file=sys.stderr)
            sys.exit(3)
        rel_id = resolve_relation_id(cat, relation_options)
        if rel_id and prop_names.get('category'):
            props[prop_names['category']] = {'relation': [{'id': rel_id}]}

        remarks = []
        if row.get('remarks'):
            remarks.append(str(row['remarks']))
        if row.get('type') == 'income':
            remarks.append('Transaction marked as income/refund offset')
        if remarks and prop_names.get('remarks'):
            props[prop_names['remarks']] = {'rich_text': [{'text': {'content': '; '.join(remarks)[:1900]}}]}

        resp = notion_post('/pages', {'parent': {'database_id': target_db}, 'properties': props})
        if resp.status_code != 200:
            print('Create page failed', resp.status_code, resp.text, file=sys.stderr)
            sys.exit(3)

    print(json.dumps({'ok': True, 'count': len(signed_rows)}))


if __name__ == '__main__':
    main()

File: src/integrations/notion/submitRows.ts
import { normalizeRows, parseAmount } from "../../domain/expenseSigns";
import type { ExpenseRow } from "../../types";
import { NotionClient } from "./notionClient";

interface NotionPage {
  id: string;
  properties?: Record<string, unknown>;
}

interface PropertyNames {
  title?: string;
  amount?: string;
  date?: string;
  category?: string;
  remarks?: string;
}

function normalizeText(value: unknown): string {
  return String(value ?? "").toLowerCase().replace(/[^a-z0-9]+/g, "");
}

function extractTitleFromPage(page: NotionPage): string {
  const properties = page.properties ?? {};
  for (const prop of Object.values(properties)) {
    if (!prop || typeof prop !== "object") {
      continue;
    }
    const typed = prop as Record<string, unknown>;
    if (typed.type !== "title" || !Array.isArray(typed.title)) {
      continue;
    }
    return typed.title
      .map((chunk) => {
        if (!chunk || typeof chunk !== "object") {
          return "";
        }
        const c = chunk as Record<string, unknown>;
        if (typeof c.plain_text === "string") {
          return c.plain_text;
        }
        if (c.text && typeof c.text === "object") {
          const t = c.text as Record<string, unknown>;
          return typeof t.content === "string" ? t.content : "";
        }
        return "";
      })
      .join("")
      .trim();
  }
  return "";
}

function normalizeDate(value: unknown): string | null {
  if (!value) {
    return null;
  }
  const raw = String(value).trim();
  if (!raw) {
    return null;
  }

  const now = new Date();
  const currentYear = now.getFullYear();
  const currentMonth = now.getMonth() + 1;
  const currentDay = now.getDate();

  const toIso = (year: number, month: number, day: number): string | null => {
    if (!Number.isInteger(year) || !Number.isInteger(month) || !Number.isInteger(day)) {
      return null;
    }
    if (month < 1 || month > 12 || day < 1 || day > 31) {
      return null;
    }
    const dt = new Date(year, month - 1, day);
    if (
      Number.isNaN(dt.getTime()) ||
      dt.getFullYear() !== year ||
      dt.getMonth() + 1 !== month ||
      dt.getDate() !== day
    ) {
      return null;
    }
    return `${year.toString().padStart(4, "0")}-${month.toString().padStart(2, "0")}-${day.toString().padStart(2, "0")}`;
  };

  const isoMatch = raw.match(/^(\d{4})-(\d{2})-(\d{2})(?:[T ].*)?$/);
  if (isoMatch) {
    return toIso(Number(isoMatch[1]), Number(isoMatch[2]), Number(isoMatch[3]));
  }

  const dmy = raw.match(/^(\d{1,2})[\/-](\d{1,2})[\/-](\d{4})$/);
  if (dmy) {
    return toIso(Number(dmy[3]), Number(dmy[2]), Number(dmy[1]));
  }

  const dmyNoYear = raw.match(/^(\d{1,2})[\/-](\d{1,2})$/);
  if (dmyNoYear) {
    return toIso(currentYear, Number(dmyNoYear[2]), Number(dmyNoYear[1]));
  }

  const ym = raw.match(/^(\d{4})-(\d{1,2})$/);
  if (ym) {
    return toIso(Number(ym[1]), Number(ym[2]), currentDay);
  }

  const my = raw.match(/^(\d{1,2})[\/-](\d{4})$/);
  if (my) {
    return toIso(Number(my[2]), Number(my[1]), currentDay);
  }

  const dayOnly = raw.match(/^(\d{1,2})$/);
  if (dayOnly) {
    return toIso(currentYear, currentMonth, Number(dayOnly[1]));
  }

  const parsed = new Date(raw);
  if (Number.isNaN(parsed.getTime())) {
    return null;
  }

  // JS Date parser can default missing years to 2001 (e.g. "3/4").
  // If no explicit 4-digit year was provided, use current year.
  const hasExplicitYear = /\b\d{4}\b/.test(raw);
  const year = hasExplicitYear ? parsed.getFullYear() : currentYear;
  const month = parsed.getMonth() + 1;
  const day = parsed.getDate();
  return toIso(year, month, day);
}

function getPropertyNames(schema: Record<string, unknown>): PropertyNames {
  const props = (schema.properties ?? {}) as Record<string, Record<string, unknown>>;
  const found: PropertyNames = {};

  for (const [name, prop] of Object.entries(props)) {
    const type = String(prop.type ?? "");
    const lowered = name.toLowerCase();
    if (type === "title" && (!found.title || lowered === "item")) {
      found.title = name;
    } else if (type === "number" && (!found.amount || lowered === "amount")) {
      found.amount = name;
    } else if (type === "date" && (!found.date || lowered === "date")) {
      found.date = name;
    } else if (type === "relation" && (!found.category || lowered === "category")) {
      found.category = name;
    } else if (type === "rich_text" && (!found.remarks || lowered === "remarks")) {
      found.remarks = name;
    }
  }

  return found;
}

interface RelationOption {
  id: string;
  title: string;
}

async function getRelationOptions(client: NotionClient, schema: Record<string, unknown>, names: PropertyNames): Promise<RelationOption[]> {
  if (!names.category) {
    return [];
  }
  const properties = (schema.properties ?? {}) as Record<string, Record<string, unknown>>;
  const relationInfo = properties[names.category]?.relation as Record<string, unknown> | undefined;
  const relationDbId = relationInfo?.database_id;
  if (typeof relationDbId !== "string") {
    return [];
  }

  const pages = await client.queryDatabase(relationDbId);
  const options: RelationOption[] = [];
  for (const page of pages as NotionPage[]) {
    const title = extractTitleFromPage(page);
    if (title) {
      options.push({ id: page.id, title });
    }
  }
  return options;
}

function resolveRelationId(rawCategory: unknown, relationOptions: RelationOption[]): string | null {
  if (!rawCategory || relationOptions.length === 0) {
    return null;
  }

  const raw = String(rawCategory).trim();
  const rawLower = raw.toLowerCase();
  const titleMap = new Map<string, string>(relationOptions.map((o) => [normalizeText(o.title), o.id]));

  const aliases: Array<[string, string]> = [
    ["transport", "Transport"], ["bus", "Transport"], ["mrt", "Transport"], ["grab", "Transport"],
    ["taxi", "Transport"], ["train", "Transport"], ["food", "Food"], ["grocer", "Food"],
    ["shopping", "Lifestyle"], ["lifestyle", "Lifestyle"], ["software", "Lifestyle"],
    ["subscription", "Lifestyle"], ["invest", "Investments"], ["goal", "Goals"],
    ["saving", "Savings"], ["tith", "Tithing"], ["buffer", "Buffer"]
  ];

  for (const [needle, canonical] of aliases) {
    if (rawLower.includes(needle)) {
      const match = titleMap.get(normalizeText(canonical));
      if (match) {
        return match;
      }
    }
  }

  const normalizedRaw = normalizeText(raw);
  if (titleMap.has(normalizedRaw)) {
    return titleMap.get(normalizedRaw) ?? null;
  }

  const loose = relationOptions.find((o) => rawLower.includes(o.title.toLowerCase()) || o.title.toLowerCase().includes(rawLower));
  return loose?.id ?? null;
}

export async function fetchRelationOptionTitles(client: NotionClient, targetDbId: string): Promise<string[]> {
  const schema = (await client.getDatabase(targetDbId)) as Record<string, unknown>;
  const names = getPropertyNames(schema);
  const options = await getRelationOptions(client, schema, names);
  return options.map((o) => o.title);
}

export async function submitRowsToNotion(client: NotionClient, targetDbId: string, rows: ExpenseRow[]): Promise<{ ok: true; count: number }> {
  const schema = (await client.getDatabase(targetDbId)) as Record<string, unknown>;
  const names = getPropertyNames(schema);
  if (!names.title) {
    throw new Error("Could not find title property in target database schema");
  }

  const relationOptions = await getRelationOptions(client, schema, names);
  const signedRows = normalizeRows(rows);

  for (const row of signedRows) {
    const item = row.item || "Unknown";
    const amount = parseAmount(row.amount);
    const isoDate = normalizeDate(row.date);
    if (row.date && !isoDate) {
      throw new Error(`Invalid date after normalization: ${String(row.date)}`);
    }

    const relationId = resolveRelationId(row.category, relationOptions);
    const properties: Record<string, unknown> = {
      [names.title]: { title: [{ text: { content: item } }] }
    };

    if (amount !== null && names.amount) {
      properties[names.amount] = { number: amount };
    }
    if (isoDate && names.date) {
      properties[names.date] = { date: { start: isoDate } };
    }
    if (relationId && names.category) {
      properties[names.category] = { relation: [{ id: relationId }] };
    }

    const remarks: string[] = [];
    if (row.remarks) {
      remarks.push(String(row.remarks));
    }
    if (row.type === "income") {
      remarks.push("Transaction marked as income/refund offset");
    }
    if (remarks.length > 0 && names.remarks) {
      properties[names.remarks] = { rich_text: [{ text: { content: remarks.join("; ").slice(0, 1900) } }] };
    }

    await client.createPage({ databaseId: targetDbId, properties });
  }

  return { ok: true, count: signedRows.length };
}
